Add device time for supported broadlink devices

eifinger opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The Broadlink thermostat devices have a large time drift of up to several minutes per week. This platform allows to sync  time to the device.

Especially after DST changes this saves users from having to walk to each device and click buttons.

As an example this can also be run with an automation every night.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/32712

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

eifinger committed:
Add service broadlink.sync_time for climate devices

PeteRager reviewed (commented):
In the tests, consider validating the parameters (hour, minute, day, etc) get passed to the device call correctly.  May require mocking dt_util.now().

eifinger commented:
> In the tests, consider validating the parameters (hour, minute, day, etc) get passed to the device call correctly. May require mocking dt_util.now().

Also considered this but decided that the simple logic does not warrant such a complex test.

felipediel commented:
Thanks @eifinger!

Have you thought about using a [datetime entity](https://developers.home-assistant.io/docs/core/entity/datetime) to show/update the date/time?

eifinger commented:
> Thanks @eifinger!
> 
> Have you thought about using a [datetime entity](https://developers.home-assistant.io/docs/core/entity/datetime) to show/update the date/time?

I thought about supplying a datetime to the service but could not think of a valid usecase why it would be different from `now()`. Can you think of a usecase for doing that?

Also nice idea to use a datetime entity for that 👍

eifinger commented:
> Thanks @eifinger!
> 
> Have you thought about using a [datetime entity](https://developers.home-assistant.io/docs/core/entity/datetime) to show/update the date/time?

For showing the current time of the device I would need to find out how it works. Moreover I wanted to stay simple.

eifinger commented:
> > Thanks @eifinger!
> > Have you thought about using a [datetime entity](https://developers.home-assistant.io/docs/core/entity/datetime) to show/update the date/time?
> 
> For showing the current time of the device I would need to find out how it works. Moreover I wanted to stay simple.

For future reference: https://github.com/mjg59/python-broadlink/blob/66744707f517593de8276ed550cecae31c204c57/broadlink/climate.py#L83-L109

felipediel commented:
Thanks @eifinger! Here's a [datetime implementation](
https://github.com/home-assistant/core/blob/dae56222e90868ce3514012cee9a08ba6b649580/homeassistant/components/knx/datetime.py#L87-L104) for reference. It is straightforward, we basically need:

- a property named native_value that returns the datetime fetched by the update coordinador (in our case it is stored in self._coordinator.data)
```python3
    @property
    def native_value(self) -> datetime | None:
        """Return the latest value."""
        if (data := self._coordinator.data) is None:
            return None

        now = dt.now()
        device_weekday = data["dayofweek"] - 1
        this_weekday = now.weekday()

        if device_weekday != this_weekday:
            days_diff = this_weekday - device_weekday
            if days_diff < 0:
                days_diff += 7
            now -= timedelta(days=days_diff)

        return now.replace(
            hour=data["hour"],
            minute=data["min"],
            second=data["sec"],
        )
```

- a method named async_set_value() that receives a datetime as parameter and sends it to the device
```python3
    async def async_set_value(self, value: datetime) -> None:
        """Change the value."""
        await self._device.async_request(
            self._device.api.set_time,
            hour=value.hour,
            minute=value.minute,
            second=value.second,
            day=value.weekday() + 1,
        )
```

I don't see any practical scenarios beyond exposing the device date/time and adjusting it manually or through automations, but I would still prefer to use the native Home Assistant component. Leveraging native components offers several benefits, such as a user-friendly interface and the potential for services and automations to be reused seemlessly, without the user having to learn the verbatim of our integration.

eifinger committed:
Use datetime instead

eifinger commented:
> Thanks @eifinger! Here's a [datetime implementation](https://github.com/home-assistant/core/blob/dae56222e90868ce3514012cee9a08ba6b649580/homeassistant/components/knx/datetime.py#L87-L104) for reference. It is straightforward, we basically need:
> 
> * a property named native_value that returns the datetime fetched by the update coordinador (in our case it is stored in self._coordinator.data)
> 
> ```python
>     @property
>     def native_value(self) -> datetime | None:
>         """Return the latest value."""
>         if (data := self._coordinator.data) is None:
>             return None
> 
>         now = dt.now()
>         device_weekday = data["dayofweek"] - 1
>         this_weekday = now.weekday()
> 
>         if device_weekday != this_weekday:
>             days_diff = this_weekday - device_weekday
>             if days_diff < 0:
>                 days_diff += 7
>             now -= timedelta(days=days_diff)
> 
>         return now.replace(
>             hour=data["hour"],
>             minute=data["min"],
>             second=data["sec"],
>         )
> ```
> 
> * a method named async_set_value() that receives a datetime as parameter and sends it to the device
> 
> ```python
>     async def async_set_value(self, value: datetime) -> None:
>         """Change the value."""
>         await self._device.async_request(
>             self._device.api.set_time,
>             hour=value.hour,
>             minute=value.minute,
>             second=value.second,
>             day=value.weekday() + 1,
>         )
> ```
> 
> I don't see any practical scenarios beyond exposing the device date/time and adjusting it manually or through automations, but I would still prefer to use the native Home Assistant component. Leveraging native components offers several benefits, such as a user-friendly interface and the potential for services and automations to be reused seemlessly, without the user having to learn the verbatim of our integration.

Removed the service and implemented the datetime platform instead.

felipediel commented on homeassistant/components/broadlink/datetime.py:
```diff
@@ -0,0 +1,77 @@
+        super().__init__(device)
+
+        self._attr_unique_id = f"{device.unique_id}-datetime"
+        self._update_internal_state()
```
We don't need to update the internal state here, the coordinator will push the first state to the entity.

felipediel commented on homeassistant/components/broadlink/datetime.py:
```diff
@@ -0,0 +1,77 @@
+        elif (data := self._coordinator.data) is None or "dayofweek" not in data:
+            self._attr_native_value = None
+        else:
+            now = dt_util.now()
+            device_weekday = data["dayofweek"] - 1
+            this_weekday = now.weekday()
+
+            if device_weekday != this_weekday:
+                days_diff = this_weekday - device_weekday
+                if days_diff < 0:
+                    days_diff += 7
+                now -= timedelta(days=days_diff)
+
+            self._attr_native_value = now.replace(
+                hour=data["hour"],
+                minute=data["min"],
+                second=data["sec"],
+            )
```
We can move this block to _update_state(), checking if data is None before and returning in case it is.

felipediel commented on homeassistant/components/broadlink/datetime.py:
```diff
@@ -0,0 +1,77 @@
+            second=value.second,
+            day=value.weekday() + 1,
+        )
+        self._update_internal_state(value)
```
Here we can assign self._attr_native_value = value directly instead of calling this method.

felipediel commented on homeassistant/components/broadlink/datetime.py:
```diff
@@ -0,0 +1,77 @@
+        self._attr_unique_id = f"{device.unique_id}-datetime"
+        self._update_internal_state()
+
+    def _update_internal_state(self, value: datetime | None = None) -> None:
```
Let's try to eliminate this function.

felipediel reviewed (changes requested):
Looks great @eifinger! A few changes and we'll be all set 🚀

eifinger commented:
> Looks great @eifinger! A few changes and we'll be all set 🚀

Just to be sure we are aligned, I had the version you are asking for before implementing the current one. With the current implementation the new state will directly be reflected. With the requested changes the new state takes up to 30s to be reflected (experience from testing).

I am all for removing that ugly code. But only if we all agree that a up to 30s delay is acceptable for our users?

eifinger committed:
Remove uneeded method

felipediel reviewed (approved):
LGTM 👍🚀

eifinger committed:
Use async_set_time_zone in tests

eifinger committed:
Test datetime update from device

joostlek commented on homeassistant/components/broadlink/datetime.py:
```diff
@@ -0,0 +1,71 @@
+    _attr_has_entity_name = True
+    _attr_native_value: datetime | None = None
+
+    def __init__(self, device) -> None:
```
Can we type this

joostlek commented on homeassistant/components/broadlink/datetime.py:
```diff
@@ -0,0 +1,71 @@
+        """Initialize the sensor."""
+        super().__init__(device)
+
+        self._attr_unique_id = f"{device.unique_id}-datetime"
```
datetime is a bit broad. maybe `device-datetime` or something would at least explain more what the entity does. (not that anyone ever looks at the unique_ids, but if you are going to add more datetimes in the future, having one that is literally called `datetime` is strange

joostlek commented on homeassistant/components/broadlink/datetime.py:
```diff
@@ -0,0 +1,71 @@
+
+        self._attr_unique_id = f"{device.unique_id}-datetime"
+
+    def _update_state(self, data) -> None:
```
types

joostlek commented on homeassistant/components/broadlink/datetime.py:
```diff
@@ -0,0 +1,71 @@
+            self._attr_native_value = None
+        else:
+            now = dt_util.now()
+            device_weekday = data["dayofweek"] - 1
```
Did they write their backend in javscript or something

eifinger replied:
maybe matlab?^^

joostlek commented on tests/components/broadlink/test_datetime.py:
```diff
@@ -0,0 +1,65 @@
+    state = hass.states.get(datetime.entity_id)
+    assert state.state == "2024-04-30T03:04:05+00:00"
+    assert mock_setup.api.set_time.call_count == 1
+    assert mock_setup.api.set_time.call_args.kwargs["hour"] == 3
+    assert mock_setup.api.set_time.call_args.kwargs["minute"] == 4
+    assert mock_setup.api.set_time.call_args.kwargs["second"] == 5
+    assert mock_setup.api.set_time.call_args.kwargs["day"] == 2
```
Let's just compare as dict

joostlek commented on tests/components/broadlink/test_datetime.py:
```diff
@@ -0,0 +1,65 @@
+    await hass.services.async_call(
+        DATETIME_DOMAIN,
+        SERVICE_SET_VALUE,
+        {"entity_id": datetime.entity_id, ATTR_DATETIME: "2024-04-30T03:04:05+00:00"},
```
```suggestion
        {ATTR_ENTITY_ID: datetime.entity_id, ATTR_DATETIME: "2024-04-30T03:04:05+00:00"},
```

joostlek commented on tests/components/broadlink/test_datetime.py:
```diff
@@ -0,0 +1,65 @@
+from . import get_device
+
+
+async def test_datetime(
```
Can we also test this part?
```py
            if device_weekday != this_weekday:
                days_diff = this_weekday - device_weekday
                if days_diff < 0:
                    days_diff += 7
                now -= timedelta(days=days_diff)
```

eifinger commented:
Sorry. Accidentally amended the commit instead of creating a new one

eifinger committed:
Split in time and select platform

joostlek commented:
as discussed, should be 2 PRs, setting to draft
